src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTag.cs

```diff
@@ -0,0 +1,164 @@
+                FontSize = 0.75 * format.FontRenderingEmSize,
+                FontStyle = FontStyles.Normal,
+                Foreground = format.ForegroundBrush,
+                Padding = new Thickness(left: 2, top: 0, right: 2, bottom: 0),
+            };
+
+            var id = new Run(_diagnostic.Id);
+            Hyperlink? link = null;
+
+            if (_diagnostic.HelpLink is null)
+            {
+                block.Inlines.Add(id);
+            }
+            else
+            {
+                link = new Hyperlink(id)
+                {
+                    NavigateUri = new Uri(_diagnostic.HelpLink)
+                };
+
+                link.RequestNavigate += HandleRequestNavigate;
+                block.Inlines.Add(link);
+            }
+
+            block.Inlines.Add(": " + _diagnostic.Message);
+
+            var lineHeight = Math.Floor(format.Typeface.FontFamily.LineSpacing * block.FontSize);
+            var image = new CrispImage
+            {
+                Moniker = GetMoniker(),
+                MaxHeight = lineHeight,
+                Margin = new Thickness(1, 0, 5, 0)
+            };
+
+            var border = new Border
+            {
+                BorderBrush = format.BackgroundBrush,
+                BorderThickness = new Thickness(1),
+                Background = Brushes.Transparent,
+                Child = new StackPanel
+                {
+                    Height = lineHeight,
+                    Orientation = Orientation.Horizontal,
+                    Children = { image, block }
+                },
+                CornerRadius = new CornerRadius(2),
+                // Highlighting lines are 2px buffer. So shift us up by one from the bottom so we feel centered between them.
+                Margin = new Thickness(10, top: 0, 0, bottom: 1),
+                Padding = new Thickness(1)
+            };
+
+            // This is used as a workaround to the moniker issues in blue theme
+            var editorBackground = (Color)_editorFormatMap.GetProperties("TextView Background")["BackgroundColor"];
+            ImageThemingUtilities.SetImageBackgroundColor(border, editorBackground);
+
+            border.Measure(new Size(double.PositiveInfinity, double.PositiveInfinity));
+            view.ViewportWidthChanged += ViewportWidthChangedHandler;
+
+            return new GraphicsResult(border, dispose:
+                () =>
+                {
+                    if (link is not null)
+                    {
+                        link.RequestNavigate -= HandleRequestNavigate;
+                    }
+
+                    view.ViewportWidthChanged -= ViewportWidthChangedHandler;
+                });
+
+            void ViewportWidthChangedHandler(object s, EventArgs e)
+            {
+                if (Location is InlineDiagnosticsLocations.PlacedAtEndOfEditor)
+                {
+                    Canvas.SetLeft(border, view.ViewportWidth - border.DesiredSize.Width);
```

Comment:
this updates if the viewport width changes.  what updates things if the user just scrolls left/right?

Reply:
That's a good point, I need to make it so it still stays at the end of the editor.